prov/gni/src/gnix_util.c

```diff
@@ -109,12 +123,96 @@ void _gnix_alps_cleanup(void)
 	if (gnix_app_peCpus)
 		free(gnix_app_peCpus);
 
-	alps_init = 0;
+	alps_init = false;
 
 	alps_app_lli_unlock();
 }
 
-static int gnix_alps_init(void)
+void _gnix_app_cleanup(void)
+{
+	if (alps_init) {
+		__gnix_alps_cleanup();
+	} else if (ccm_init) {
+		__gnix_ccm_cleanup();
+	}
+}
+
+static int __gnix_ccm_init(void)
+{
+	int rc, fd;
+	FILE *f;
+	char ccm_nodelist[PATH_MAX];
+	const char *home, *jobid;
+	ccm_alps_info_t info;
+	uint32_t num_nids = 0;
+
+	GNIX_DEBUG(FI_LOG_FABRIC, "Reading job info file %s",
+		   CCM_ALPS_INFO_FILE);
+
+	fd = open(CCM_ALPS_INFO_FILE, O_RDONLY);
+	if (fd < 0) {
+		return -FI_EIO;
+	}
+
+	rc = read(fd, &info, sizeof(ccm_alps_info_t));
+	if (rc != sizeof(ccm_alps_info_t))
+		return -FI_EIO;
+
+	gnix_app_ptag = info.ptag;
+	gnix_app_cookie = info.cookie;
+
+	close(fd);
+	GNIX_DEBUG(FI_LOG_FABRIC, "Ptag=0x%x, cookie=0x%x",
+		   gnix_app_ptag, gnix_app_cookie);
+
+	home = getenv("HOME");
+	jobid = getenv("PBS_JOBID");
+	snprintf(ccm_nodelist, PATH_MAX, "%s/%s%s", home ? home : ".",
+		 CCM_NODELIST_FN, jobid ? jobid : "sdb");
+	f = fopen(ccm_nodelist, "r");
+	if (f) {
+		char nid0[128];
+		char next_nid[128];
+
+		rc = fscanf(f, "%s\n", nid0);
+		/* assume this one worked, error case is same */
+		num_nids++;
```

Comment:
Are you looking for the number of instances on a node?
Or the number of nodes?
To me a nid means node.